### PR TITLE
Use ObjectMethodExecutor for hub method invocations

### PR DESCRIPTION
**Pull request by BrennanConroy**
#144

### CODE REVIEW

File: src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+using System.ComponentModel;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.SignalR.Internal
+{
+    class ObjectMethodExecutor
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs]: internal

File: src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+using System.ComponentModel;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.SignalR.Internal
+{
+    internal class ObjectMethodExecutor
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs]: Can we add tests for this class? I don't think our tests cover all the cases like static vs. instance methods, Task vs Task<T> vs, void return types, default parameter values etc.

Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs]: Here's the original tests https://github.com/aspnet/Mvc/blob/760c8f38678118734399c58c2dac981ea6e47046/test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ObjectMethodExecutorTest.cs
Handles most scenarios, do you want to copy them?

Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs]: I guess so - we will rev this code independently.

File: src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+                var valueCast = Expression.Convert(valueObj, paramInfo.ParameterType);
+
+                // valueCast is "(Ti) parameters[i]"
+                parameters.Add(valueCast);
+            }
+
+            // Call method
+            MethodCallExpression methodCall;
+
+            if (!methodInfo.IsStatic)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs]: FYI: This is the stuff I added to the class to allow for static methods

Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs]: The very reason why we need to have our own tests

File: src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs (hunk 2 above)
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs]: Unless we're going to actually make this a common library/shared package with tests, we need to test it on our end as well.

File: src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+        private static readonly MethodInfo _convertOfTMethod =
+            typeof(ObjectMethodExecutor).GetRuntimeMethods().Single(methodInfo => methodInfo.Name == nameof(ObjectMethodExecutor.Convert));
+
+        private ObjectMethodExecutor(MethodInfo methodInfo, TypeInfo targetTypeInfo)
+        {
+            if (methodInfo == null)
+            {
+                throw new ArgumentNullException(nameof(methodInfo));
+            }
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs]: nit: new line here

File: src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs (hunk 2 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs]: @rynowak should we make this a shared component?

**Comment by rynowak** [comment]
Yeah it would be dope to make this shared. For instance we're going to be adding F# support to it soon. Y'all will want that as well.

**Comment by anurse** [comment]
I think a shared-source package would be all we need. It's not quite worthy of it's own assembly.

**Comment by BrennanConroy** [comment]
Couple minor things, they reference IActionResult and we need static method support and they don't

File: src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs (hunk 3 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/Internal/ObjectMethodExecutor.cs]: It's possible we want to disallow this.

File: src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs [hunk 5]
```diff
@@ -262,21 +263,24 @@ public class HubEndPoint<THub, TClient> : EndPoint, IInvocationBinder where THub
...
+                    object result = null;
+                    if (methodExecutor.IsMethodAsync)
                     {
-                        await resultTask;
-                        if (methodInfo.ReturnType.GetTypeInfo().IsGenericType)
+                        if (methodExecutor.TaskGenericType == null)
                         {
-                            var property = resultTask.GetType().GetProperty("Result");
-                            invocationResult.Result = property?.GetValue(resultTask);
+                            await (Task)methodExecutor.Execute(hub, invocationDescriptor.Arguments);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs]: Is this pretty weird no? Why doesn't ExecuteAsync work here? If not that seems like worthy refactoring after this is merged.

Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs]: ExecuteAsync calls TaskOfTActionExecutorAsync which assumes you have a Task<T> return type. I would be okay with refactoring it.

**Pull request merged**